File: src/engine/render/Renderer.ts
import type { GameConfig } from "../core/Config.js";
import { UniformBuffer } from "./UniformBuffer.js";
import { ShaderProgram } from "./ShaderProgram.js";
import { Texture2DArray } from "./Texture2DArray.js";
import { FrustumCuller } from "./FrustumCuller.js";
import { ChunkMesh } from "./ChunkMesh.js";
import { chunkVertexShaderSource, chunkFragmentShaderSource } from "./shaders/chunk.js";
import { skyVertexShaderSource, skyFragmentShaderSource } from "./shaders/sky.js";
import type { CameraView } from "./CameraView.js";
import { TexturePipeline } from "../assets/TexturePipeline.js";
import type { BlockRegistry } from "../../world/BlockRegistry.js";
import type { World } from "../../world/World.js";

const CAMERA_BLOCK_FLOATS = 80;
const TIME_BLOCK_FLOATS = 8;

export class Renderer {
  private readonly vertexStrideFloats = 10;
  private readonly chunkShader: ShaderProgram;
  private readonly skyShader: ShaderProgram;
  private readonly cameraUbo: UniformBuffer;
  private readonly timeUbo: UniformBuffer;
  private readonly textures: Texture2DArray;
  private readonly texturePipeline = new TexturePipeline();
  private readonly frustum = new FrustumCuller();
  private readonly meshes = new Map<string, ChunkMesh>();
  private readonly cameraBlock = new Float32Array(CAMERA_BLOCK_FLOATS);
  private readonly frustumMin = new Float32Array(3);
  private readonly frustumMax = new Float32Array(3);
  private readonly skyVao: WebGLVertexArrayObject;
  private readonly skyVbo: WebGLBuffer;

  constructor(
    private readonly gl: WebGL2RenderingContext,
    blocks: BlockRegistry,
    private readonly config: GameConfig,
  ) {
    this.chunkShader = new ShaderProgram(gl, chunkVertexShaderSource, chunkFragmentShaderSource);
    this.skyShader = new ShaderProgram(gl, skyVertexShaderSource, skyFragmentShaderSource);
    this.chunkShader.bindUniformBlock("CameraBlock", 0);
    this.chunkShader.bindUniformBlock("TimeBlock", 2);
    this.skyShader.bindUniformBlock("CameraBlock", 0);
    this.skyShader.bindUniformBlock("TimeBlock", 2);
    this.cameraUbo = new UniformBuffer(gl, 0, CAMERA_BLOCK_FLOATS * 4);
    this.timeUbo = new UniformBuffer(gl, 2, TIME_BLOCK_FLOATS * 4);
    this.textures = new Texture2DArray(gl, 16, 16, config.textureArrayLayers);
    this.seedTextureArray(blocks);
    const skyVao = gl.createVertexArray();
    const skyVbo = gl.createBuffer();
    if (!skyVao || !skyVbo) throw new Error("Failed to create sky resources");
    this.skyVao = skyVao;
    this.skyVbo = skyVbo;
    gl.bindVertexArray(this.skyVao);
    gl.bindBuffer(gl.ARRAY_BUFFER, this.skyVbo);
    gl.bufferData(gl.ARRAY_BUFFER, new Float32Array([-1, -1, 3, -1, -1, 3]), gl.STATIC_DRAW);
    gl.enableVertexAttribArray(0);
    gl.vertexAttribPointer(0, 2, gl.FLOAT, false, 0, 0);
    gl.bindVertexArray(null);
    gl.bindBuffer(gl.ARRAY_BUFFER, null);

    gl.enable(gl.DEPTH_TEST);
    gl.disable(gl.CULL_FACE);
    gl.enable(gl.BLEND);
    gl.blendFunc(gl.SRC_ALPHA, gl.ONE_MINUS_SRC_ALPHA);
  }

  get glContext(): WebGL2RenderingContext {
    return this.gl;
  }

  getTimeUbo(): UniformBuffer {
    return this.timeUbo;
  }

  bindBlockTextures(unit = 0): void {
    this.textures.bind(unit);
  }

  resizeCanvasToDisplaySize(): number {
    this.resizeCanvas();
    return this.gl.canvas.width / this.gl.canvas.height;
  }

  render(world: World, camera: CameraView, timeSeconds: number, daylightFactor: number): void {

    this.syncChunks(world);

    this.gl.viewport(0, 0, this.gl.canvas.width, this.gl.canvas.height);
    const skyR = 0.08 + 0.5 * daylightFactor;
    const skyG = 0.1 + 0.64 * daylightFactor;
    const skyB = 0.16 + 0.74 * daylightFactor;
    this.gl.clearColor(0, 0, 0, 1);
    this.gl.clear(this.gl.COLOR_BUFFER_BIT | this.gl.DEPTH_BUFFER_BIT);

    this.uploadCameraBlock(camera, timeSeconds, daylightFactor, skyR, skyG, skyB);
    this.renderSky();

    this.chunkShader.use();
    this.bindBlockTextures(0);
    this.gl.uniform1i(this.chunkShader.uniform("u_blockTextures"), 0);

    this.frustum.extractFrom(camera.viewProjectionMatrix);

    for (const [key, chunk] of world.entries()) {
      const mesh = this.meshes.get(key);
      if (!mesh || mesh.indexCount === 0) continue;

      this.frustumMin[0] = chunk.chunkX * this.config.chunkSize;
      this.frustumMin[1] = 0;
      this.frustumMin[2] = chunk.chunkZ * this.config.chunkSize;
      this.frustumMax[0] = this.frustumMin[0] + this.config.chunkSize;
      this.frustumMax[1] = this.config.worldHeight;
      this.frustumMax[2] = this.frustumMin[2] + this.config.chunkSize;

      if (!this.frustum.testAABB(this.frustumMin, this.frustumMax)) continue;

      this.gl.uniform3f(
        this.chunkShader.uniform("u_chunkTranslation"),
        chunk.chunkX * this.config.chunkSize,
        0,
        chunk.chunkZ * this.config.chunkSize,
      );
      mesh.draw();
    }
  }

  dispose(): void {
    for (const mesh of this.meshes.values()) mesh.dispose();
    this.meshes.clear();
    this.textures.dispose();
    this.cameraUbo.dispose();
    this.timeUbo.dispose();
    this.gl.deleteBuffer(this.skyVbo);
    this.gl.deleteVertexArray(this.skyVao);
    this.chunkShader.dispose();
    this.skyShader.dispose();
  }

  private syncChunks(world: World): void {
    for (const [key, mesh] of this.meshes) {
      if (!world.hasChunkKey(key)) {
        mesh.dispose();
        this.meshes.delete(key);
      }
    }

    for (const [key, chunk] of world.entries()) {
      if (chunk.state !== "meshReady") continue;
      if (chunk.indexCount === 0 || chunk.vertexCount === 0) {
        world.markUploaded(chunk);
        continue;
      }

      let mesh = this.meshes.get(key);
      if (!mesh) {
        mesh = new ChunkMesh(this.gl);
        this.meshes.set(key, mesh);
      }

      const slot = world.getChunkSlot(chunk);
      const vertexView = slot.vertices.subarray(0, chunk.vertexCount * this.vertexStrideFloats);
      const indexView = slot.indices.subarray(0, chunk.indexCount);
      mesh.upload(chunk, vertexView, indexView, this.vertexStrideFloats);
      world.markUploaded(chunk);
    }
  }

  private uploadCameraBlock(
    camera: CameraView,
    timeSeconds: number,
    daylightFactor: number,
    skyR: number,
    skyG: number,
    skyB: number,
  ): void {
    this.cameraBlock.set(camera.projectionMatrix, 0);
    this.cameraBlock.set(camera.viewMatrix, 16);
    this.cameraBlock.set(camera.viewProjectionMatrix, 32);
    this.cameraBlock.set(camera.inverseViewProjectionMatrix, 48);
    this.cameraBlock[64] = camera.position[0];
    this.cameraBlock[65] = camera.position[1];
    this.cameraBlock[66] = camera.position[2];
    this.cameraBlock[67] = timeSeconds;
    this.cameraBlock[68] = skyR * (0.55 + 0.45 * daylightFactor);
    this.cameraBlock[69] = skyG * (0.55 + 0.45 * daylightFactor);
    this.cameraBlock[70] = skyB * (0.6 + 0.4 * daylightFactor);
    this.cameraBlock[71] = this.config.renderDistance * this.config.chunkSize * 1.8;
    this.cameraBlock[72] = camera.right[0];
    this.cameraBlock[73] = camera.right[1];
    this.cameraBlock[74] = camera.right[2];
    this.cameraBlock[75] = 0;
    this.cameraBlock[76] = camera.up[0];
    this.cameraBlock[77] = camera.up[1];
    this.cameraBlock[78] = camera.up[2];
    this.cameraBlock[79] = 0;
    this.cameraUbo.upload(this.cameraBlock);
  }

  private renderSky(): void {
    this.gl.depthMask(false);
    this.gl.disable(this.gl.CULL_FACE);
    this.skyShader.use();
    this.gl.bindVertexArray(this.skyVao);
    this.gl.drawArrays(this.gl.TRIANGLES, 0, 3);
    this.gl.bindVertexArray(null);
    this.gl.depthMask(true);
  }

  private seedTextureArray(blocks: BlockRegistry): void {
    for (const layer of this.texturePipeline.createLayerArray(this.config.textureArrayLayers)) {
      this.textures.uploadLayer(layer.layer, layer.pixels, layer.width, layer.height);
    }
    this.textures.generateMipmaps();
    void blocks;
  }

  private resizeCanvas(): void {
    const canvas = this.gl.canvas as HTMLCanvasElement;
    const width = Math.max(1, Math.floor(canvas.clientWidth * window.devicePixelRatio));
    const height = Math.max(1, Math.floor(canvas.clientHeight * window.devicePixelRatio));
    if (canvas.width === width && canvas.height === height) return;
    canvas.width = width;
    canvas.height = height;
  }
}

File: src/engine/render/Renderer.cpp
#include "Renderer.hpp"
#include "shaders/ShaderSources.hpp"
#include "world/daynight/DayNightCycle.hpp"
#include "engine/assets/AssetManager.hpp"
#include <array>
#include <cmath>
#include <algorithm>
#include <cstring>
namespace voxel {

Renderer::Renderer(GLFWwindow* window, BlockRegistry& blocks, const GameConfig& config,
                   ChunkMeshAllocator& meshAllocator)
  : m_window(window), m_config(config),
    m_chunkShader(shaders::chunkVertex, shaders::chunkFragment),
    m_skyShader(shaders::skyVertex, shaders::skyFragment),
    m_cameraUbo(0, CAMERA_BLOCK_FLOATS * sizeof(float)),
    m_timeUbo(2, TIME_BLOCK_FLOATS * sizeof(float)),
    m_textures(16, 16, std::max(1, AssetManager::get().getTextureLayerCount())),
    m_meshAllocator(meshAllocator),
    m_indirectBatcher([&]{
        int32_t poolCap = (config.renderDistance * 2 + 1) * (config.renderDistance * 2 + 1) + 8;
        return std::make_unique<IndirectBatcher>(poolCap);
    }()),
    m_chunkSyncer(m_meshAllocator, *m_indirectBatcher, config),
    m_drawDispatcher(m_chunkShader, m_skyShader, m_textures, *m_indirectBatcher, m_masterVao, m_skyVao)
{
  (void)blocks;

  m_chunkShader.bindUniformBlock("CameraBlock", 0);
  m_chunkShader.bindUniformBlock("TimeBlock", 2);
  m_skyShader.bindUniformBlock("CameraBlock", 0);
  m_skyShader.bindUniformBlock("TimeBlock", 2);

  // Create sky fullscreen triangle
  gl::GenVertexArrays(1, &m_skyVao);
  gl::GenBuffers(1, &m_skyVbo);
  gl::BindVertexArray(m_skyVao);
  gl::BindBuffer(GL_ARRAY_BUFFER, m_skyVbo);
  float skyVerts[] = {-1.0f, -1.0f, 3.0f, -1.0f, -1.0f, 3.0f};
  gl::BufferData(GL_ARRAY_BUFFER, sizeof(skyVerts), skyVerts, GL_STATIC_DRAW);
  gl::EnableVertexAttribArray(0);
  gl::VertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE, 0, nullptr);
  gl::BindVertexArray(0);
  gl::BindBuffer(GL_ARRAY_BUFFER, 0);

  gl::Enable(GL_DEPTH_TEST);
  gl::Disable(GL_CULL_FACE);
  // Blending is toggled per-pass in DrawDispatcher: off for opaque, on for transparent.
  gl::Disable(GL_BLEND);
  gl::BlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

  gl::GenVertexArrays(1, &m_masterVao);
  gl::BindVertexArray(m_masterVao);
  gl::BindBuffer(GL_ARRAY_BUFFER, m_meshAllocator.vboBuffer());
  gl::BindBuffer(GL_ELEMENT_ARRAY_BUFFER, m_meshAllocator.iboBuffer());

  size_t strideBytes = m_config.vertexStrideFloats * sizeof(float);
  gl::EnableVertexAttribArray(0);
  gl::VertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, strideBytes, reinterpret_cast<void*>(0));
  gl::EnableVertexAttribArray(1);
  gl::VertexAttribPointer(1, 3, GL_FLOAT, GL_FALSE, strideBytes, reinterpret_cast<void*>(12));
  gl::EnableVertexAttribArray(2);
  gl::VertexAttribPointer(2, 2, GL_FLOAT, GL_FALSE, strideBytes, reinterpret_cast<void*>(24));
  gl::EnableVertexAttribArray(3);
  gl::VertexAttribPointer(3, 1, GL_FLOAT, GL_FALSE, strideBytes, reinterpret_cast<void*>(32));
  gl::EnableVertexAttribArray(4);
  gl::VertexAttribPointer(4, 1, GL_FLOAT, GL_FALSE, strideBytes, reinterpret_cast<void*>(36));
  gl::BindVertexArray(0);

  seedTextureArray();
  updateFramebufferSize();
}

Renderer::~Renderer() {
  dispose();
}

auto Renderer::updateFramebufferSize() -> float {
  glfwGetFramebufferSize(m_window, &m_fbWidth, &m_fbHeight);
  m_fbWidth = std::max(1, m_fbWidth);
  m_fbHeight = std::max(1, m_fbHeight);
  return static_cast<float>(m_fbWidth) / static_cast<float>(m_fbHeight);
}

void Renderer::render(World& world, const CameraView& camera,
                       float timeSeconds, float daylightFactor,
                       float ambientIntensity, float normalizedTimeOfDay) {
  m_chunkSyncer.sync(world);
  m_frustum.extractFrom(camera.viewProjectionMatrix);

  gl::Viewport(0, 0, m_fbWidth, m_fbHeight);

  float skyR = 0.08f + 0.5f * daylightFactor;
  float skyG = 0.1f + 0.64f * daylightFactor;
  float skyB = 0.16f + 0.74f * daylightFactor;
  gl::ClearColor(skyR, skyG, skyB, 1.0f);
  gl::Clear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

  uploadCameraBlock(camera, timeSeconds, daylightFactor, skyR, skyG, skyB);

  // Upload time block
  {
    float dayFac = daylightFactor;
    glm::vec3 sunDir = daynight::computeSunDirection(timeSeconds);
    glm::vec3 sunCol = daynight::computeSunColor(timeSeconds);
    float sunIntensity = dayFac; // sun contribution scales with daylight

    // std140 layout (80 bytes = 20 floats):
    //   0: u_timeElapsed
    //   4: u_daylight
    //   8: u_sunIntensity
    //  12: u_ambientIntensity
    //  16: u_sunDir.x
    //  20: u_sunDir.y
    //  24: u_sunDir.z
    //  28: padding
    //  32: u_timeOfDay
    //  36: padding
    //  40: padding
    //  44: padding
    //  48: u_sunColor.r
    //  52: u_sunColor.g
    //  56: u_sunColor.b
    //  60: padding
    //  64: u_pad
    //  68: padding
    //  72: padding
    //  76: padding
    std::array<float, TIME_BLOCK_FLOATS> timeData{};
    timeData[0] = timeSeconds;          // u_timeElapsed
    timeData[1] = dayFac;               // u_daylight (0=night, 1=day)
    timeData[2] = sunIntensity;         // u_sunIntensity
    timeData[3] = ambientIntensity;     // u_ambientIntensity
    timeData[4] = sunDir.x;             // u_sunDir.x
    timeData[5] = sunDir.y;             // u_sunDir.y
    timeData[6] = sunDir.z;             // u_sunDir.z
    timeData[8] = normalizedTimeOfDay;   // u_timeOfDay (0-1)
    timeData[12] = sunCol.r;            // u_sunColor.r
    timeData[13] = sunCol.g;            // u_sunColor.g
    timeData[14] = sunCol.b;            // u_sunColor.b
    timeData[16] = 0.0f;                // u_pad
    m_timeUbo.upload(timeData.data(), sizeof(timeData));
  }

  m_drawDispatcher.renderSky();
  m_drawDispatcher.renderChunks(m_frustum);
}

void Renderer::dispose() {
  if (m_skyVao) gl::DeleteVertexArrays(1, &m_skyVao);
  if (m_skyVbo) gl::DeleteBuffers(1, &m_skyVbo);
  m_skyVao = 0;
  m_skyVbo = 0;

  if (m_masterVao) gl::DeleteVertexArrays(1, &m_masterVao);
  m_masterVao = 0;
  m_indirectBatcher.reset();
}

// ---- Private ----

void Renderer::uploadCameraBlock(const CameraView& camera, float timeSeconds,
                                  float daylightFactor, float skyR, float skyG, float skyB) {
  auto* cb = m_cameraBlock.data();

  // u_proj (mat4) at [0..15]
  std::memcpy(cb, &camera.projectionMatrix[0], 16 * sizeof(float));
  // u_view (mat4) at [16..31]
  std::memcpy(cb + 16, &camera.viewMatrix[0], 16 * sizeof(float));
  // u_projView (mat4) at [32..47]
  std::memcpy(cb + 32, &camera.viewProjectionMatrix[0], 16 * sizeof(float));
  // u_invProjView (mat4) at [48..63]
  std::memcpy(cb + 48, &camera.inverseViewProjectionMatrix[0], 16 * sizeof(float));
  // u_camTime (vec4) at [64..67]
  cb[64] = camera.position.x;
  cb[65] = camera.position.y;
  cb[66] = camera.position.z;
  cb[67] = timeSeconds;
  // u_fogColor (vec4) at [68..71]
  cb[68] = skyR * (0.55f + 0.45f * daylightFactor);
  cb[69] = skyG * (0.55f + 0.45f * daylightFactor);
  cb[70] = skyB * (0.6f + 0.4f * daylightFactor);
  cb[71] = static_cast<float>(m_config.renderDistance) * m_config.chunkSize * 1.8f;
  // u_camRight (vec4) at [72..75]
  cb[72] = camera.right.x;
  cb[73] = camera.right.y;
  cb[74] = camera.right.z;
  cb[75] = 0.0f;
  // u_camUp (vec4) at [76..79]
  cb[76] = camera.up.x;
  cb[77] = camera.up.y;
  cb[78] = camera.up.z;
  cb[79] = 0.0f;

  m_cameraUbo.upload(cb, CAMERA_BLOCK_FLOATS * sizeof(float));
}

void Renderer::seedTextureArray() {
  int32_t layers = AssetManager::get().getTextureLayerCount();
  if (layers == 0) return;

  const auto& pixelData = AssetManager::get().getTextureData();

  if (m_textures.isHighBitDepth()) {
    // Single upload for all layers — far less driver overhead than N calls.
    m_textures.uploadAllLayers(pixelData.data(), 16, 16, layers);
  } else {
    // Fallback: convert all layers to 8-bit at once, then single upload.
    const size_t totalPixels = static_cast<size_t>(layers) * 16 * 16 * 4;
    std::vector<uint8_t> fallbackData(totalPixels);
    const uint16_t* src = pixelData.data();
    for (size_t i = 0; i < totalPixels; ++i) {
      fallbackData[i] = static_cast<uint8_t>(src[i] >> 8);
    }
    m_textures.uploadAllLayers8(fallbackData.data(), 16, 16, layers);
  }
}

} // namespace voxel
